### PR TITLE
change buildspec check behavior

### PR DESCRIPTION
**Pull request by shahzebsiddiqui**
discover_buildspec will return None if no file is found, this can be a result of invalid file,
 no files during directory traversal. We will print a message and 
not raise an error. discover_buildspec is called once per --buildspec 
argument so we only add return of discover_buildspec if a valid 
file is returned. BuildspecParser will not expect invalid file since all 
checks are done in advance, so we refactor regression test.

There is no need for BuildspecParser.load method instead we move
``_validate_global`` and ``_validate`` to ``__init__``.
Furthermore, we move **self.recipe** to load schema in __init__ rather than
invoking in _validate_global. Class variable should be declared in
__init__.
Also buildspec is a required argument to BuildspecParser, it not
possible to have this as an optional argument

### CODE REVIEW

**Comment by shahzebsiddiqui** [comment]
@vsoch here is my proposal which could be merged to https://github.com/HPC-buildtest/buildtest-framework/pull/277 or merge https://github.com/HPC-buildtest/buildtest-framework/pull/277 as is and i will sync this branch and send this for upstream.

I have done some tests to ensure buildtest will terminate early 

Here are some examples

1. Invalid file

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b xas
Please provide an absolute or relative path to a directory file from your present working directory or /u/users/ssi29/.buildtest/site
There are no config files to process.
```

2. Invalid extension


```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b README.rst
README.rst does not end in file extension .yaml or .yml
There are no config files to process.
```

3. Directory with no files retrieved

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b docs
There are no Buildspecs to process for argument: docs
There are no config files to process.
```

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b docs -b dir1 -b di2
There are no Buildspecs to process for argument: docs
Please provide an absolute or relative path to a directory file from your present working directory or /u/users/ssi29/.buildtest/site
Please provide an absolute or relative path to a directory file from your present working directory or /u/users/ssi29/.buildtest/site
There are no config files to process.
```

Since ``discover_buildspecs`` is called once per ``--buildspec`` if buildtest discovers files after processing each argument it will proceed. It won't terminate immediately unless all arguments have been processed. 

For example here is an example where buildtest will process files with one valid argument

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b docs -b examples/benchmarks/
There are no Buildspecs to process for argument: docs

            Discovered Buildspecs

/mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml



Buildspec Name                 SubTest                        Status                         Buildspec Path
________________________________________________________________________________________________________________________
stream                         stream_uniprocess_c            PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml
stream                         stream_openmp_c                PASSED                         /mxg-hpc/users/ssi29/buildtest-framework/examples/benchmarks/stream.yml


============================================================
                        Test summary
============================================================
Executed 2 tests
Passed Tests: 2/2 Percentage: 100.000%
Failed Tests: 0/2 Percentage: 0.000%

```

This check is done after exclusion so if files are discovered but after exclusion there are no files, buildtest will stop immediately.

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b docs -b examples/benchmarks/ -x examples/benchmarks/
There are no Buildspecs to process for argument: docs
There are no Buildspec files to process.
```

Furthermore, I have cloned a repo using ``buildtest get`` and tried to build by directory and it searches as expected.

```
(buildtest-framework) ssi29@ag-mxg-hulk090> buildtest build -b github.com/HPC-buildtest/buildtest-stampede2/tests/system

            Discovered Buildspecs

/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/timezone.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/account_alias.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/selinux.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/shared_fs.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/ulimits.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/dns.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/bzip2_compression.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/check_login.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/module.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/mount.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/idev.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/systemd.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/disk_usage.yml
/u/users/ssi29/.buildtest/site/github.com/HPC-buildtest/buildtest-stampede2/tests/system/kernel.yml
```

One thing i dont like is having to type a path from ``github.com`` all the way to the desired located, since i have many test i don't want to fire all of them at once.

File: buildtest/buildsystem/base.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def __init__(self, buildspec):
 
         self.logger = logging.getLogger(__name__)
 
-        self.recipe = None
+        self.buildspec = os.path.abspath(buildspec)
```
Comment (vsoch) [inline buildtest/buildsystem/base.py:64]: This will be an error if buildspec is given as None.

File: buildtest/buildsystem/base.py [hunk 2]
```diff
@@ -70,27 +70,7 @@ def __init__(self, buildspec):
...
-        self.buildspec = os.path.abspath(buildspec)
-
-        if not is_file(self.buildspec):
-            sys.exit("Buildspec File: %s does not exist." % self.buildspec)
-
-        elif is_dir(self.buildspec):
-            sys.exit(
-                "Please provide a file path (not a directory path) to a Buildspec file."
-            )
+        self.recipe = load_recipe(self.buildspec)
```
Comment (vsoch) [inline buildtest/buildsystem/base.py:73]: I absolutely don't support this. The load function is there explicitly for error handling, and removing it with the argument "but the client is doing checking it won't ever get here" is not good enough.

**Comment by vsoch** [comment]
What exactly is your proposal? To remove error checking from the BuildtestParser class? The point isn't that it wouldn't be caught from the client, the point is that the instantiation of the parser, which could be done by any user of buildtest as a module, needs error checking. This absolutely cannot be merged, it's not logical, and is bad practice to remove error checking for a class.

**Comment by shahzebsiddiqui** [comment]
@vsoch okay i am going to close this PR since this change doesn't make sense.

**Pull request closed without merging**